### DISCUSSION TITLE
Multi value delimiter remove spaces between keywords

### DISCUSSION THREAD

**Discussion by thewebartisan7** [spatie/laravel-query-builder#514] [Q&A]
As explained here https://docs.spatie.be/laravel-query-builder/v2/advanced-usage/multi-value-delimiter/ I am using multiple keywords separated by comma, so there is one input for search where users can add keywords separated by comma.

All works perfectly but if users add with spaces after comma (example `keyword1, keyword2, keyword3`) then query become like:

```
select count(*) as aggregate from `users` where (LOWER(`name`) LIKE '%keyword1%' or LOWER(`name`) LIKE '% keyword2%' or LOWER(`name`) LIKE '% keyword3%') and `users`.`deleted_at` is null
```

If I change array value delimiter'   to comma + space ( ', ' ) then works fine. But if users then search without space after comma (example `keyword1,keyword2,keyword3`) then query become:

```
select count(*) as aggregate from `users` where LOWER(`name`) LIKE '%keyword1,keyword2,keyword3%' and `users`.`deleted_at` is null
```

There is a workaround to solve this without adding extra code to clean filter? Like checking if exist comma + space and then replace with just comma... but seem annoying to maintein.

In a such great package I suppose there is something but can't find anything in docs. 

thank You!

**Comment by dominikb** [comment]
Hi, thanks for opening this discussion.

I don't think that normalizing and sanitizing user input is in the scope of this package.

One way you could work around this issue is by preparing the user input before it is passed into this package, either by using a middleware on the server-side or by preparing the input before submitting the request from the frontend:

1. Attempt to replace all occurrences of whitespace+comma with just a comma
[regexr Example](https://regexr.com/5aqqn)
```php
$preparedInput = preg_replace("/\s*,\s*/", ",", $subject)
```

2. Try to use a regex to grep valid inputs
[regexr Example](https://regexr.com/5aqtk)
```php
$pattern = '/([^,]+)(?:[ ,]*)?/';

preg_match_all($pattern, $subject, $matches);

$matches = array_map('trim', $matches[1]);
```

Hope this helps!

**Comment by thewebartisan7** [comment]
Thanks for your reply. Make sense to not put sanitization inside this package, however I see this more like a correction, like making lowercase keywords in sql query. Because if value delimiter make multiple conditions, it's unlikely that someone would start or end a search with empty space. Like Laravel have middleware for trim spaces, it make sense to trim space also in this case, since delimiter it's like creating a multiple values of the same input. I hope that I explain well myself.

Thanks for sharing solution. The best is second. The first don't solve fancy input with only multiple comma like

`keyword1, keyword2, , , , , ,keyword3`  

This return 

`keyword1,keyword2,,,,,,keyword3`

So the final code become:

```php

            if(request()->has('filter.q')) {
                $pattern = '/([^,]+)(?:[ ,]*)?/';
                preg_match_all($pattern, request()->filter['q'], $matches);
                $matches = array_map('trim', $matches[1]);
                request()->merge([
                    'filter' => [
                        'q' => implode(',', $matches)
                    ]
                ]);
            }

```

I will put this in a middleware, but I think would not be bad to add this in package, only in this case when delimiter are used. A bit of correction is done also for include if I understand correctly digging a bit inside the code.

Best
